cvmfs/revision.py

```diff
@@ -0,0 +1,226 @@
+            return self.catalog
+
+    def __init__(self, revision):
+        root_catalog = revision.retrieve_root_catalog()
+        self.revision    = revision
+        self.catalog_stack = collections.deque()
+        wrapper            = self._CatalogWrapper(self.revision)
+        wrapper.catalog    = root_catalog
+        self._push_catalog_wrapper(wrapper)
+
+    def __iter__(self):
+        return self
+
+    def next(self):
+        if not self._has_more():
+            raise StopIteration()
+        catalog = self._pop_catalog()
+        self._push_nested_catalogs(catalog)
+        return catalog
+
+    def _has_more(self):
+        return len(self.catalog_stack) > 0
+
+    def _push_nested_catalogs(self, catalog):
+        for nested_reference in catalog.list_nested():
+            wrapper = self._CatalogWrapper(self.revision)
+            wrapper.catalog_reference = nested_reference
+            self._push_catalog_wrapper(wrapper)
+
+    def _push_catalog_wrapper(self, catalog):
+        self.catalog_stack.append(catalog)
+
+    def _pop_catalog(self):
+        wrapper = self.catalog_stack.pop()
+        return wrapper.get_catalog()
+
+
+class Revision:
+    """ Wrapper around a CVMFS Repository revision.
+    A Revision is a concrete instantiation in time of the Repository. It
+    represents the concrete status of the repository in a certain period of
+    time. Revision data is contained in the so-called Tags, which are stored in
+    the History database.
+    """
+
+    def __init__(self, repository, tag):
+        self.repository = repository
+        self._tag = tag
+
+    def __str__(self):
+        return '<Revision ' + self.revision_number \
+               + ' - ' + self.root_hash + '>'
+
+    @property
+    def revision_number(self):
+        return self._tag.revision
+
+    @property
+    def name(self):
+        return self._tag.name
+
+    @property
+    def timestamp(self):
+        return self._tag.timestamp
+
+    @property
+    def root_hash(self):
+        return self._tag.hash
+
+    def retrieve_catalog(self, catalog_hash):
+        """ Download and open a catalog that belongs to this revision """
+        return self.repository.retrieve_catalog(catalog_hash)
+
+    def retrieve_root_catalog(self):
+        return self.retrieve_catalog(self.root_hash)
+
+    def catalogs(self):
+        return CatalogTreeIterator(self)
+
+    def retrieve_catalog_for_path(self, needle_path):
+        """
+        Recursively walk down the Catalogs and find the best fit for a path
+        """
+        clg = self.retrieve_root_catalog()
+        while True:
+            new_nested_reference = clg.find_nested_for_path(needle_path)
+            if new_nested_reference is None:
+                break
+            nested_reference = new_nested_reference
+            clg = self.retrieve_catalog(nested_reference.hash)
+        return clg
+
+    def _opened_catalogs_for_path(self, path):
```

Comment:
We need to rethink this method, I believe. With multiple `Revision` objects it might happen that we have multiple `Catalog`s with the same `root_prefix`. Hence, this method might pick up the wrong `Catalog`.